<a href="https://colab.research.google.com/github/DimaKav/DS-Unit-2-Sprint-3-Advanced-Regression/blob/master/DS_Unit_2_Sprint_Challenge_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 2 Sprint Challenge 3

## Logistic Regression and Beyond

In this sprint challenge you will fit a logistic regression modeling the probability of an adult having an income above 50K. The dataset is available at UCI:

https://archive.ics.uci.edu/ml/datasets/adult

Your goal is to:

1. Load, validate, and clean/prepare the data.
2. Fit a logistic regression model
3. Answer questions based on the results (as well as a few extra questions about the other modules)

Don't let the perfect be the enemy of the good! Manage your time, and make sure to get to all parts. If you get stuck wrestling with the data, simplify it (if necessary, drop features or rows) so you're able to move on. If you have time at the end, you can go back and try to fix/improve.

### Hints

It has a variety of features - some are continuous, but many are categorical. You may find [pandas.get_dummies](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html) (a method to one-hot encode) helpful!

The features have dramatically different ranges. You may find [sklearn.preprocessing.minmax_scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.minmax_scale.html#sklearn.preprocessing.minmax_scale) helpful!

## Part 1 - Load, validate, and prepare data

The data is available at: https://archive.ics.uci.edu/ml/datasets/adult

Load it, name the columns, and make sure that you've loaded the data successfully. Note that missing values for categorical variables can essentially be considered another category ("unknown"), and may not need to be dropped.

You should also prepare the data for logistic regression - one-hot encode categorical features as appropriate.

In [0]:
# TODO - your work
from sklearn.linear_model import Ridge
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.model_selection import KFold, cross_val_score
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression

In [0]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data',header=None)

In [0]:
df.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [0]:
df.columns = ['age','workclass','fnlwgt','education','education-num',
              'marital-status','occupation','relationship','race',
              'sex','capital-gain','cpital-loss','hours-per-week',
              'native-country','income']

In [0]:
df.head(20)

age          workclass  fnlwgt      education  education-num  \
0    39          State-gov   77516      Bachelors             13   
1    50   Self-emp-not-inc   83311      Bachelors             13   
2    38            Private  215646        HS-grad              9   
3    53            Private  234721           11th              7   
4    28            Private  338409      Bachelors             13   
5    37            Private  284582        Masters             14   
6    49            Private  160187            9th              5   
7    52   Self-emp-not-inc  209642        HS-grad              9   
8    31            Private   45781        Masters             14   
9    42            Private  159449      Bachelors             13   
10   37            Private  280464   Some-college             10   
11   30          State-gov  141297      Bachelors             13   
12   23            Private  122272      Bachelors             13   
13   32            Private  205019     Assoc-acdm             12   
14   40            Private  121772      Assoc-voc             11   
15   34            Private  245487        7th-8th              4   
16   25   Self-emp-not-inc  176756        HS-grad              9   
17   32            Private  186824        HS-grad              9   
18   38            Private   28887           11th              7   
19   43   Self-emp-not-inc  292175        Masters             14   

            marital-status          occupation    relationship  \
0            Never-married        Adm-clerical   Not-in-family   
1       Married-civ-spouse     Exec-managerial         Husband   
2                 Divorced   Handlers-cleaners   Not-in-family   
3       Married-civ-spouse   Handlers-cleaners         Husband   
4       Married-civ-spouse      Prof-specialty            Wife   
5       Married-civ-spouse     Exec-managerial            Wife   
6    Married-spouse-absent       Other-service   Not-in-family   
7       Married-civ-spouse     Exec-managerial         Husband   
8            Never-married      Prof-specialty   Not-in-family   
9       Married-civ-spouse     Exec-managerial         Husband   
10      Married-civ-spouse     Exec-managerial         Husband   
11      Married-civ-spouse      Prof-specialty         Husband   
12           Never-married        Adm-clerical       Own-child   
13           Never-married               Sales   Not-in-family   
14      Married-civ-spouse        Craft-repair         Husband   
15      Married-civ-spouse    Transport-moving         Husband   
16           Never-married     Farming-fishing       Own-child   
17           Never-married   Machine-op-inspct       Unmarried   
18      Married-civ-spouse               Sales         Husband   
19                Divorced     Exec-managerial       Unmarried   

                   race      sex  capital-gain  cpital-loss  hours-per-week  \
0                 White     Male          2174            0              40   
1                 White     Male             0            0              13   
2                 White     Male             0            0              40   
3                 Black     Male             0            0              40   
4                 Black   Female             0            0              40   
5                 White   Female             0            0              40   
6                 Black   Female             0            0              16   
7                 White     Male             0            0              45   
8                 White   Female         14084            0              50   
9                 White     Male          5178            0              40   
10                Black     Male             0            0              80   
11   Asian-Pac-Islander     Male             0            0              40   
12                White   Female             0            0              30   
13                Black     Male             0            0              50   
14 

In [0]:
df.shape

(32561, 15)

In [0]:
df.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
cpital-loss       0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [0]:
df.describe(include='all')

age workclass        fnlwgt education  education-num  \
count   32561.000000     32561  3.256100e+04     32561   32561.000000   
unique           NaN         9           NaN        16            NaN   
top              NaN   Private           NaN   HS-grad            NaN   
freq             NaN     22696           NaN     10501            NaN   
mean       38.581647       NaN  1.897784e+05       NaN      10.080679   
std        13.640433       NaN  1.055500e+05       NaN       2.572720   
min        17.000000       NaN  1.228500e+04       NaN       1.000000   
25%        28.000000       NaN  1.178270e+05       NaN       9.000000   
50%        37.000000       NaN  1.783560e+05       NaN      10.000000   
75%        48.000000       NaN  2.370510e+05       NaN      12.000000   
max        90.000000       NaN  1.484705e+06       NaN      16.000000   

             marital-status       occupation relationship    race    sex  \
count                 32561            32561        32561   32561  32561   
unique                    7               15            6       5      2   
top      Married-civ-spouse   Prof-specialty      Husband   White   Male   
freq                  14976             4140        13193   27816  21790   
mean                    NaN              NaN          NaN     NaN    NaN   
std                     NaN              NaN          NaN     NaN    NaN   
min                     NaN              NaN          NaN     NaN    NaN   
25%                     NaN              NaN          NaN     NaN    NaN   
50%                     NaN              NaN          NaN     NaN    NaN   
75%                     NaN              NaN          NaN     NaN    NaN   
max                     NaN              NaN          NaN     NaN    NaN   

        capital-gain   cpital-loss  hours-per-week  native-country  income  
count   32561.000000  32561.000000    32561.000000           32561   32561  
unique           NaN           NaN             NaN              42       2  
top              NaN           NaN             NaN   United-States   <=50K  
freq             NaN           NaN             NaN           29170   24720  
mean     1077.648844     87.303830       40.437456             NaN     NaN  
std      7385.292085    402.960219       12.347429             NaN     NaN  
min         0.000000      0.000000        1.000000             NaN     NaN  
25%         0.000000      0.000000       40.000000             NaN     NaN  
50%         0.000000      0.000000       40.000000             NaN     NaN  
75%         0.000000      0.000000       45.000000             NaN     NaN  
max     99999.000000   4356.000000       99.000000             NaN     NaN

In [0]:
df['workclass'].value_counts()

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64

In [0]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
cpital-loss        int64
hours-per-week     int64
native-country    object
income            object
dtype: object

In [0]:
df['occupation'].value_counts()

 Prof-specialty       4140
 Craft-repair         4099
 Exec-managerial      4066
 Adm-clerical         3770
 Sales                3650
 Other-service        3295
 Machine-op-inspct    2002
 ?                    1843
 Transport-moving     1597
 Handlers-cleaners    1370
 Farming-fishing       994
 Tech-support          928
 Protective-serv       649
 Priv-house-serv       149
 Armed-Forces            9
Name: occupation, dtype: int64

In [0]:
df['native-country'].value_counts()

 United-States                 29170
 Mexico                          643
 ?                               583
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Guatemala                        64
 Japan                            62
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 

In [0]:
df['income'].value_counts()

 <=50K    24720
 >50K      7841
Name: income, dtype: int64

In [0]:
df['income'] = pd.factorize(df['income'])[0]
df['income'].value_counts()

0    24720
1     7841
Name: income, dtype: int64

In [0]:
# One hot encode objects
df_1 = pd.get_dummies(df, df.select_dtypes('object').columns)

## Part 2 - Fit and present a Logistic Regression

Your data should now be in a state to fit a logistic regression. Use scikit-learn, define your `X` (independent variable) and `y`, and fit a model.

Then, present results - display coefficients in as interpretible a way as you can (hint - scaling the numeric features will help, as it will at least make coefficients more comparable to each other). If you find it helpful for interpretation, you can also generate predictions for cases (like our 5 year old rich kid on the Titanic) or make visualizations - but the goal is your exploration to be able to answer the question, not any particular plot (i.e. don't worry about polishing it).

It is *optional* to use `train_test_split` or validate your model more generally - that is not the core focus for this week. So, it is suggested you focus on fitting a model first, and if you have time at the end you can do further validation.

In [0]:
df_1.corr()['income'].sort_values()

marital-status_ Never-married           -0.318440
relationship_ Own-child                 -0.228532
sex_ Female                             -0.215980
relationship_ Not-in-family             -0.188497
occupation_ Other-service               -0.156348
relationship_ Unmarried                 -0.142857
education_ HS-grad                      -0.131189
marital-status_ Divorced                -0.126995
occupation_ Adm-clerical                -0.089986
race_ Black                             -0.089089
occupation_ Handlers-cleaners           -0.087265
education_ 11th                         -0.085864
relationship_ Other-relative            -0.083716
occupation_ ?                           -0.078583
workclass_ Private                      -0.078528
workclass_ ?                            -0.078198
marital-status_ Separated               -0.074386
education_ 10th                         -0.070038
occupation_ Machine-op-inspct           -0.069401
marital-status_ Widowed                 -0.064381


In [0]:
# Standardize
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df_1.loc[:, df_1.columns != 'income']))
df_scaled['income'] = df_1['income']
df_scaled.head()

/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype uint8, int64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/anaconda3/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype uint8, int64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


0         1         2         3        4         5        6  \
0  0.030671 -1.063611  1.134739  0.148453 -0.21666 -0.035429 -0.24445   
1  0.837109 -1.008707  1.134739 -0.145920 -0.21666 -2.222153 -0.24445   
2 -0.042642  0.245079 -0.420060 -0.145920 -0.21666 -0.035429 -0.24445   
3  1.057047  0.425801 -1.197459 -0.145920 -0.21666 -0.035429 -0.24445   
4 -0.775768  1.408176  1.134739 -0.145920 -0.21666 -0.035429 -0.24445   

          7         8         9   ...          99       100       101  \
0 -0.174295 -0.262097 -0.014664   ...   -0.059274 -0.019201 -0.049628   
1 -0.174295 -0.262097 -0.014664   ...   -0.059274 -0.019201 -0.049628   
2 -0.174295 -0.262097 -0.014664   ...   -0.059274 -0.019201 -0.049628   
3 -0.174295 -0.262097 -0.014664   ...   -0.059274 -0.019201 -0.049628   
4 -0.174295 -0.262097 -0.014664   ...   -0.059274 -0.019201 -0.049628   

        102       103       104       105       106       107  income  
0 -0.039607 -0.023518 -0.024163  0.340954 -0.045408 -0.022173       0  
1 -0.039607 -0.023518 -0.024163  0.340954 -0.045408 -0.022173       0  
2 -0.039607 -0.023518 -0.024163  0.340954 -0.045408 -0.022173       0  
3 -0.039607 -0.023518 -0.024163  0.340954 -0.045408 -0.022173       0  
4 -0.039607 -0.023518 -0.024163 -2.932948 -0.045408 -0.022173       0  

[5 rows x 109 columns]

In [0]:
X = df_scaled.loc[:, df_scaled.columns != 'income']
y = df_scaled['income']
logreg = LogisticRegression(solver='lbfgs').fit(X,y)
logreg.score(X,y)

0.8533828813611376

In [0]:
cols = df_1.loc[:, df_1.columns != 'income'].columns

In [0]:
df_c = pd.DataFrame(logreg.coef_.T, index=cols, columns=['coefficients'])
df_c.head()

coefficients
age                0.347843
fnlwgt             0.074610
education-num      0.369452
capital-gain       2.343789
cpital-loss        0.260586

## Part 3 - Analysis, Interpretation, and Questions

### Based on your above model, answer the following questions

1. What are 3 features positively correlated with income above 50k?

2. What are 3 features negatively correlated with income above 50k?

3. Overall, how well does the model explain the data and what insights do you derive from it?

*These answers count* - that is, make sure to spend some time on them, connecting to your analysis above. There is no single right answer, but as long as you support your reasoning with evidence you are on the right track.

Note - scikit-learn logistic regression does *not* automatically perform a hypothesis test on coefficients. That is OK - if you scale the data they are more comparable in weight.

### Match the following situation descriptions with the model most appropriate to addressing them

In addition to logistic regression, a number of other approaches were covered this week. Pair them with the situations they are most appropriate for, and briefly explain why.

Situations:
1. You are given data on academic performance of primary school students, and asked to fit a model to help predict "at-risk" students who are likely to receive the bottom tier of grades.

Quantile Regression. We can get the quantile of the at-risk students and fit a quantile regression model to see what differentiates them from students in a quantile directly above.

2. You are studying tech companies and their patterns in releasing new products, and would like to be able to model and predict when a new product is likely to be launched.

Survival analysis. Survival analysis is great for analyzing data where the outcome variable is time until the occurrence of an event. Using survival analysis, we can see the release patterns of new products under various constraints and then pick one that models the characteristiscs that we're looking for.

3. You are working on modeling expected plant size and yield with a laboratory that is able to capture fantastically detailed physical data about plants, but only of a few dozen plants at a time.

Ridge regression. Ridge regression allows you to regularize coefficients of many variables, so it's better a option for high demensional data.

Approaches:
1. Ridge Regression
2. Quantile Regression
3. Survival Analysis

### Top 3 positively and negatively correlated features from Pearson

From Pearson correlation:

Positively:

* marital-status_ Married-civ-spouse 
* relationship_ Husband       
* education-num                    

Negatively:

* marital-status_ Never-married           
* relationship_ Own-child                 
* sex_ Female    

## 1. Top 5 positively correlated features from regression:

These coefficients tell us how much the dependent variable is expected to increase when the independent variable increases by one (when income is above 50k).

In [0]:
df_c.sort_values(by='coefficients', ascending=False)[0:5]

coefficients
capital-gain                            2.343789
marital-status_ Married-civ-spouse      0.756693
education-num                           0.369452
hours-per-week                          0.366557
age                                     0.347843

These coefficients tell us how much the dependent variable is expected to decrease when the independent variable increases by one (when income is above 50k)

In [0]:
df_c.sort_values(by='coefficients', ascending=False)[-5:]

coefficients
occupation_ Other-service         -0.255739
occupation_ Priv-house-serv       -0.275013
relationship_ Own-child           -0.298034
education_ Preschool              -0.474479
marital-status_ Never-married     -0.536948

## 3. How well does the model explain the data?

The model explains the data well and everythng seems to make intuitive sense. To make the model more accurate, it would be appropriate to check for multicolinearity between predictor variables and account it if any exists. It's impressive how much capital gains, being married, and number of hours per week positively contribute to having an income over 50k. Capital gain of people who make over 50k are 2.3 times as much. It's also interesting how having no education, never being married, and being a priv house serv negatively contribute to it.